walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,111 @@
+    Set<String> requestOptionalAttendeesSet = new HashSet<> ();
+    requestOptionalAttendeesSet.addAll(requestOptionalAttendees);
+    requestAttendeesSet.addAll(requestAttendees);
+    requestAttendeesSet.addAll(requestOptionalAttendees);
+    int requestDuration = (int) request.getDuration();
+    
+    Collection<TimeRange> allEventsTimeRanges = new ArrayList<TimeRange> ();
+    for (Iterator currEvent = events.iterator(); currEvent.hasNext(); ){ 
+      allEventsTimeRanges.add((TimeRange) ((Event) currEvent.next()).getWhen());
+    }
+
+    int startTime = 0;
+    int endTime = 0;
+    TimeRange tempTimeRange = TimeRange.fromStartEnd(startTime, endTime, true);
+    while (startTime <= 1439 && endTime <= 1439) {
+      if (checkOverlapMultipleTimeRanges(tempTimeRange, allEventsTimeRanges)) {
+        //don't add if the start=end but the timerange overlaps
+        //timerange has duration zero and should just be updated
+        if (startTime == endTime) {
+          startTime++;
+          endTime++;
+        } else {
+          //want to add if the attendees overlap and the duration is long enough
+          if (checkOverlapMultipleAttendees(tempTimeRange, requestAttendeesSet, events)) {
+            if (endTime - startTime >= requestDuration ) {
+              tempTimeRange = TimeRange.fromStartEnd(startTime, endTime-1, true);
+              availableTimeRanges.add(tempTimeRange);
+            }
+            startTime = endTime;
+          } else {
+            endTime++;
+          }
+        }
+      } else {
+        endTime++;
+      }
+      tempTimeRange = TimeRange.fromStartEnd(startTime, endTime, true);
+    }
+    
+    if (endTime - startTime > requestDuration) {
+      tempTimeRange = TimeRange.fromStartEnd(startTime, endTime-1, true);
+      availableTimeRanges.add(tempTimeRange);
+    }
+    //if the optional attendees cause there to be no available timeranges
+    //remove the optional attendees and optional events
+    if (availableTimeRanges.size() == 0 && requestOptionalAttendees.size() != 0 && requestAttendees.size() != 0) {
+      MeetingRequest requestWithoutOptionals = new MeetingRequest(requestAttendees, requestDuration);
+      Collection<Event> eventsWithoutOptionalEvents = new ArrayList<> ();
+      for (Iterator currEvent = events.iterator(); currEvent.hasNext(); ) { 
+        Event tempEvent = (Event)currEvent.next();
+        if (!checkOverlapAttendees(requestOptionalAttendeesSet, tempEvent.getAttendees())) {
+          eventsWithoutOptionalEvents.add(tempEvent);
+        }
+      }
+      return query(eventsWithoutOptionalEvents, requestWithoutOptionals);
+    } else {
+      return availableTimeRanges;
+    }
+  }
+
+  //checks if a TimeRange overlaps any of the TimeRanges in a collection
+  public boolean checkOverlapMultipleTimeRanges(TimeRange singleTimeRange, Collection <TimeRange> timeRanges) {
+    for (Iterator currTimeRange = timeRanges.iterator(); currTimeRange.hasNext(); ) { 
```

Comment:
same as above - better syntax is the default iteration:

for (TimeRange timeRange : timeRanges) {
...
}